<a href="https://colab.research.google.com/github/sanjay6131/discussion-3/blob/main/MyProjectExample/ProjectPart1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project Part # 1: CPSMA 3133 Discrete Mathematics (Dr. Jacob) 

## Sets, Relations and Functions

### Sanjay Katuwal(270080)





##Sets and Universal


I have created sets of natural numbers from 1 to 15.

My set $P$ consists of first five prime numbers and set $O$ consists of first five odd numbers and set $D$ consists of first five numbers divisible by 3. 




In [17]:
P = ['2', '3', '5', '7', '11']
O = ['1', '3', '5', '7', '9']
D = ['3', '6', '9', '12', '15']



My universal set $U$ consists of natural numbers less or equal to 15 which is a set of sets P,O,D.



Making  Universal set $U$ as Universal.

In [32]:
Universalset = [P,O,D]
U = []
for sett in Universalset:
  for numbers in sett:
    U.append(numbers)

set(U)

{'1', '11', '12', '15', '2', '3', '5', '6', '7', '9'}

## Union and Intersection

I am going to do a union and intersection with *for* loops.

In [18]:
PUO = []
for numbers in P:
  PUO.append(numbers) 
for numbers in O:
  PUO.append(numbers)

PUO = set(PUO) 

PUO 

{'1', '11', '2', '3', '5', '7', '9'}

In [19]:
DUO = []
for numbers in D:
  DUO.append(numbers) 
for numbers in O:
  DUO.append(numbers) 

DUO = set(DUO) 
DUO

{'1', '12', '15', '3', '5', '6', '7', '9'}

In [20]:
PUD = []
for numbers in P:
  PUD.append(numbers) 
for numbers in D:
  PUD.append(numbers) 

PUD = set(PUD) 
PUD

{'11', '12', '15', '2', '3', '5', '6', '7', '9'}

I used a conditional **if** for the intersection.

In [28]:
PIntO = []
for numbers in P:
  if numbers in O:
    PIntO.append(numbers)

PIntO

['3', '5', '7']

In [29]:
DIntO = []
for numbers in D:
  if numbers in O:
    DIntO.append(numbers)

DIntO

['3', '9']

In [30]:
DIntP = []
for numbers in D:
  if numbers in P:
    DIntP.append(numbers)

DIntP

['3']

In [10]:
P = ['4', '8', '10', '13', '14']

for numbers in P:
  U.append(numbers)



## Complement

I'll use the **not in** for the complement.

In [33]:
PComplement =[]

for numbers in U:
  if numbers not in P:
    PComplement.append(numbers)

PComplement

['1', '9', '6', '9', '12', '15']

In [34]:
DComplement =[]

for numbers in U:
  if numbers not in D:
    DComplement.append(numbers)

DComplement

['2', '5', '7', '11', '1', '5', '7']

In [35]:
OComplement =[]

for numbers in U:
  if numbers not in O:
    OComplement.append(numbers)

OComplement

['2', '11', '6', '12', '15']

13 is a prime number but it is still in PComplement because my set P consists of first five prime numbers and 13 comes the sixth prime number.

For the Intersections and Unions, I could not find anything interesting but Unions have almost same number of elements and intersections have almost same number of elements.

## Powersets

To make a powerset we have to use loops as it is the easiest way.  

In [36]:
def PODpowerset(s):
  powerset = [[]]
  for i in s:
    for sett in powerset:
      powerset = powerset + [sett + [i]]
  return powerset




In [37]:
PODpowerset(P)

[[],
 ['2'],
 ['3'],
 ['2', '3'],
 ['5'],
 ['2', '5'],
 ['3', '5'],
 ['2', '3', '5'],
 ['7'],
 ['2', '7'],
 ['3', '7'],
 ['2', '3', '7'],
 ['5', '7'],
 ['2', '5', '7'],
 ['3', '5', '7'],
 ['2', '3', '5', '7'],
 ['11'],
 ['2', '11'],
 ['3', '11'],
 ['2', '3', '11'],
 ['5', '11'],
 ['2', '5', '11'],
 ['3', '5', '11'],
 ['2', '3', '5', '11'],
 ['7', '11'],
 ['2', '7', '11'],
 ['3', '7', '11'],
 ['2', '3', '7', '11'],
 ['5', '7', '11'],
 ['2', '5', '7', '11'],
 ['3', '5', '7', '11'],
 ['2', '3', '5', '7', '11']]

In [38]:
PODpowerset(U)

[[],
 ['2'],
 ['3'],
 ['2', '3'],
 ['5'],
 ['2', '5'],
 ['3', '5'],
 ['2', '3', '5'],
 ['7'],
 ['2', '7'],
 ['3', '7'],
 ['2', '3', '7'],
 ['5', '7'],
 ['2', '5', '7'],
 ['3', '5', '7'],
 ['2', '3', '5', '7'],
 ['11'],
 ['2', '11'],
 ['3', '11'],
 ['2', '3', '11'],
 ['5', '11'],
 ['2', '5', '11'],
 ['3', '5', '11'],
 ['2', '3', '5', '11'],
 ['7', '11'],
 ['2', '7', '11'],
 ['3', '7', '11'],
 ['2', '3', '7', '11'],
 ['5', '7', '11'],
 ['2', '5', '7', '11'],
 ['3', '5', '7', '11'],
 ['2', '3', '5', '7', '11'],
 ['1'],
 ['2', '1'],
 ['3', '1'],
 ['2', '3', '1'],
 ['5', '1'],
 ['2', '5', '1'],
 ['3', '5', '1'],
 ['2', '3', '5', '1'],
 ['7', '1'],
 ['2', '7', '1'],
 ['3', '7', '1'],
 ['2', '3', '7', '1'],
 ['5', '7', '1'],
 ['2', '5', '7', '1'],
 ['3', '5', '7', '1'],
 ['2', '3', '5', '7', '1'],
 ['11', '1'],
 ['2', '11', '1'],
 ['3', '11', '1'],
 ['2', '3', '11', '1'],
 ['5', '11', '1'],
 ['2', '5', '11', '1'],
 ['3', '5', '11', '1'],
 ['2', '3', '5', '11', '1'],
 ['7', '11', '1'],
 ['2', 

In [86]:
print(list(powerset(P)))

[[], ['2'], ['3'], ['2', '3'], ['5'], ['2', '5'], ['3', '5'], ['2', '3', '5'], ['7'], ['2', '7'], ['3', '7'], ['2', '3', '7'], ['5', '7'], ['2', '5', '7'], ['3', '5', '7'], ['2', '3', '5', '7'], ['11'], ['2', '11'], ['3', '11'], ['2', '3', '11'], ['5', '11'], ['2', '5', '11'], ['3', '5', '11'], ['2', '3', '5', '11'], ['7', '11'], ['2', '7', '11'], ['3', '7', '11'], ['2', '3', '7', '11'], ['5', '7', '11'], ['2', '5', '7', '11'], ['3', '5', '7', '11'], ['2', '3', '5', '7', '11']]


In [85]:
print(list(powerset(Universalset)))

[[], [['2', '3', '5', '7', '11']], [['1', '3', '5', '7', '9']], [['2', '3', '5', '7', '11'], ['1', '3', '5', '7', '9']], [['3', '6', '9', '12', '15']], [['2', '3', '5', '7', '11'], ['3', '6', '9', '12', '15']], [['1', '3', '5', '7', '9'], ['3', '6', '9', '12', '15']], [['2', '3', '5', '7', '11'], ['1', '3', '5', '7', '9'], ['3', '6', '9', '12', '15']]]


## Subsets

In [50]:
def IsPSubset(P,D):
  for a in P:
    if a not in D:
      return False
  return True

IsPSubset(P,D)

False

In [84]:
IsPSubset(D,O)

False

In [83]:
IsPSubset(P,O)

False

In [75]:
IsPSubset(P,P)

True

In [76]:
IsPSubset(O,O)

True

In [77]:
IsPSubset(D,D)

True

In [78]:
IsPSubset(U,U)

True

In [79]:
IsPSubset(P,U)

True

In [81]:
IsPSubset(D,U)

True

In [82]:
IsPSubset(O,U)

True

It looks like there is not any set to be subset of other rather than itself or subset of Universal set. And there are 7 subsets in total.

Finding out number of elements in universal.

In [88]:
N= len(U)
print(N)

15


Number of elements in a powerset.

In [89]:
2**N

32768

There are 32768 numbers in poweset of  universal set.